src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -191,30 +206,225 @@ public MapLock AcquireMapLock()
+                        }
+                    }
+
+                    // This persisted node is being removed from cache. Keep it in mind in case of an update to the same
+                    // path.
+                    _pastPathHash.Set(new(key.Address, in treePath), key.Keccak);
+                }
+            }
+
+
+            public void RemovePastKeys(ConcurrentDictionary<HashAndTinyPath, Hash256?> persistedHashes, INodeStorage nodeStorage)
+            {
+                bool CanRemove(in ValueHash256 address, TinyTreePath path, in TreePath fullPath, in ValueHash256 keccak, Hash256? currentlyPersistingKeccak)
+                {
+                    // Multiple current hash that we don't keep track for simplicity. Just ignore this case.
+                    if (currentlyPersistingKeccak is null) return false;
+
+                    // The persisted hash is the same as currently persisting hash. Do nothing.
+                    if ((ValueHash256)currentlyPersistingKeccak == keccak) return false;
+
+                    // We have it in cache and it is still needed.
+                    if (TryGetValue(new TrieStoreDirtyNodesCache.Key(address, fullPath, keccak.ToCommitment()), out TrieNode node) &&
+                        !_trieStore.IsNoLongerNeeded(node)) return false;
+
+                    // We don't have it in cache, but we know it was re-committed, so if it is still needed, don't remove
+                    if (_persistedLastSeen.TryGetValue(new(address, in path, in keccak), out long commitBlock) &&
+                        !_trieStore.IsNoLongerNeeded(commitBlock)) return false;
+
+                    return true;
+                }
+
+                using (AcquireMapLock())
+                {
+                    INodeStorage.WriteBatch writeBatch = nodeStorage.StartWriteBatch();
+                    try
+                    {
+                        int round = 0;
+                        foreach (KeyValuePair<HashAndTinyPath, Hash256> keyValuePair in persistedHashes)
+                        {
+                            HashAndTinyPath key = keyValuePair.Key;
+                            if (_pastPathHash.TryGet(key, out ValueHash256 prevHash))
+                            {
+                                TreePath fullPath = key.path.ToTreePath(); // Micro op to reduce double convert
+                                if (CanRemove(key.addr, key.path, fullPath, prevHash, keyValuePair.Value))
+                                {
+                                    Metrics.RemovedNodeCount++;
+                                    Hash256? address = key.addr == default ? null : key.addr.ToCommitment();
+                                    writeBatch.Set(address, fullPath, prevHash, default, WriteFlags.DisableWAL);
+                                    round++;
+                                }
+                            }
+
+                            // Batches of 256
+                            if (round > 256)
```

Comment:
It does not seems to matter much here. The NoWAL flag should make it very quirk. Actually, I probably need to double check this.

Reply:
Another reason is that on persist before can be made so that the action of the parent runs before the children making the key insertion order mostly increasing (until it reach storage). On single thread this is a notable improvement when batch size is large. On multithread, not a reliable difference is seen. Same behavious with snap sync before where before the batch was splitted into smaller chunks, sorting the keys improve performace.

Reply:
btw are batches sorted in-memory? Can we help with sorting in any way?

Reply:
Not here. Its looping a dictionary.